cmd/gandi.go

```diff
@@ -34,32 +37,43 @@ const (
 )
 
 var (
-	resourceType = kingpin.Arg("type", "Resource type (zone, record, snapshot, domain or axfr)").Required().String()
+	apiType = kingpin.Arg("type", "API type (domain, livedns, email, billing or organisation)").Required().String()
+	resourceType = kingpin.Arg("subtype", "Resource type (record, livednsSnapshot, domain or livednsAxfr)").Required().String()
 	action       = kingpin.Arg("action", "Action (valid actions depend on the type - if you provide an erroneous action, a list of allowed actions will be displayed)").Required().String()
 	args         = kingpin.Arg("args", "Arguments to the action (valid arguments depend on the action)").Strings()
 	apiKey       = kingpin.Flag("key", "The Gandi LiveDNS API key (may be stored in the GANDI_KEY environment variable)").OverrideDefaultFromEnvar("GANDI_KEY").Short('k').String()
 	sharing_id   = kingpin.Flag("sharing_id", "The Gandi LiveDNS sharing_id (may be stored in the GANDI_SHARING_ID environment variable)").OverrideDefaultFromEnvar("GANDI_SHARING_ID").Short('i').String()
-	g            *gandi.Gandi
+	debug = kingpin.Flag("debug", "Show debug info").Bool()
+	dry_run = kingpin.Flag("dry_run", "Show debug info").Bool()
+	d *domain.Domain
+	l *livedns.LiveDNS
 )
 
 func main() {
 	kingpin.CommandLine.HelpFlag.Short('h')
 	kingpin.Parse()
-	g = gandi.New(*apiKey, *sharing_id)
-	switch *resourceType {
-	case "zone":
-		zone()
-	case "record":
-		zoneRecord()
-	case "snapshot":
-		snapshot()
+	g := gandi.Config{
+		SharingID: *sharing_id,
+		Debug:     *debug,
+		DryRun:    *dry_run,
+	}
+	d = domain.New(*apiKey, &g)
+	l = livedns.New(*apiKey, &g)
```

Comment:
@TomOnTime @tiramiseb hopefully this makes the intent with `client` a bit clearer. We now just create a config struct, and pass it directly to the constructors for the APIs; you never need import `client` or use it.

Reply:
What if I have a single product which must interact with all parts of the API? I have to instantiate a new service for each part of the same API?

Reply:
Yes, but I think you'd actually want that. Because there's a ListDomains call in both the Domains API and the LiveDNS api, and they return different objects and likely different lists of domains.

Reply:
domain api return the list of domain you own at the registry, livedns api returns the list of domain handled by the dns system.
The former is a superset of the later.

Reply:
I think it would be clearer if it was:

```
config := gandi.Config{....
...}
d = gandi_domain.New(*apiKey, config)
d = gandi_livedns.New(*apikey, config)
```

My code interacts with multiple DNS providers and if each calls themselves "domain" or "client" then I have to rename them at import.   "good package names should not require renaming." https://github.com/golang/go/wiki/CodeReviewComments#imports

Reply:
Why not a root package which would provide `NewDomain(*apiKey, config)`, `NewLiveDNS(*apiKey, config)`, etc?